eng/pipelines/libraries/enterprise/linux.yml

```diff
@@ -0,0 +1,97 @@
+trigger:
+  batch: true
+  branches:
+    include:
+    - master
+    - release/*.*
+  paths:
+    exclude:
+    - docs/*
+    - CODE-OF-CONDUCT.md
+    - CONTRIBUTING.md
+    - LICENSE.TXT
+    - PATENTS.TXT
+    - README.md
+    - SECURITY.md
+    - THIRD-PARTY-NOTICES.TXT
+    - src/installer/*
+    - src/coreclr/*
+    - eng/pipelines/coreclr/*
+    - eng/pipelines/installer/*
+    - eng/pipelines/common/*
+
+pr:
+  branches:
+    include:
+    - master
+    - release/*.*
+  paths:
+    exclude:
+    - docs/*
+    - CODE-OF-CONDUCT.md
+    - CONTRIBUTING.md
+    - LICENSE.TXT
+    - PATENTS.TXT
+    - README.md
+    - SECURITY.md
+    - THIRD-PARTY-NOTICES.TXT
+    - src/installer/*
+    - src/coreclr/*
+    - eng/pipelines/coreclr/*
+    - eng/pipelines/installer/*
+    - eng/pipelines/common/*
+
+pool:
+  vmImage: 'ubuntu-16.04'
+
+variables:
+  - template: ../variables.yml
+  - name: enterpriseTestsSetup
+    value: $(sourcesRoot)/Common/tests/System/Net/EnterpriseTests/setup
+  - name: containerRunTestsCommand
+    value: /repo/.dotnet/dotnet msbuild /t:rebuildandtest
+  - name: containerLibrariesRoot
+    value: /repo/src/libraries
+
+steps:
+- bash: |
+    cd $(enterpriseTestsSetup)
+    docker-compose build
+  displayName: Build test machine images
+  env:
+    DOTNET_RUNTIME_REPO_ROOT: $(Build.SourcesDirectory)
+
+- bash: |
+    cd $(enterpriseTestsSetup)
+    docker-compose up -d
+  displayName: Start test network and machines
+  env:
+    DOTNET_RUNTIME_REPO_ROOT: $(Build.SourcesDirectory)
+
+- bash: |
+    docker exec linuxclient bash /setup/test-webserver.sh
```

Comment:
Your use of docker exec for controlling the container is interesting, and I can probably see why you went for that approach. While it does offer flexibility and convenience, I'm wondering whether it can cause problems, particularly given the fact that it's just mounting the source files from the host filesystem. This has caused problems for me in the past, for example when building a dirty windows repo from WSL (build artifacts generated by windows confuse the linux sdk).

In the interest of reproducibility, I would recommend using a docker build which copies and builds the artifacts into a new image. The tests can then be run by calling `docker run` on the new image. Note that you can still override the `CMD` command with `docker run`, as well as issue new commands using `docker exec` while it is running. This has a few benefits, namely you can deploy the client to remote machines that do not necessarily have the source code checked out.

That being said, there is a trade-off, namely the copying of source files will certainly increase the build time. That should not be a concern from a pipelines perspective. I can see you might be motivated by setting up an interactive dev experience, but I think the same can be achieved by using, say, VsCode's RemoteSSH against a dedicated linux box.

Reply:
>I would recommend using a docker build which copies and builds the artifacts into a new image.

I considered many approaches to this including the one you mention here which is used by the stress testing app. But I prioritized the 'dev' experience and wanted the experience to be as similar as possible to running the other 'FunctionalTests' we have in the repo. I also wanted it to be fast so that developers would be inclined to use it. And this technique allows for tight 'inner loop' development. You can quickly make a change to a few source files in a library and just build that library (inside the container). And then run the tests again.

>I'm wondering whether it can cause problems, particularly given the fact that it's just mounting the source files from the host filesystem. This has caused problems for me in the past, for example when building a dirty windows repo from WSL (build artifacts generated by windows confuse the linux sdk).

This is a good point. But I think it can be addressed with documentation here to give developers some tips when running the tests on their local dev box. And the technique being used here to build the repo and run the tests by mounting from the host works perfectly fine in CI on the Azure DevOps agent's machines. Those are always using a clean VM.